metagraph/src/annotation/annotation_converters.cpp

```diff
@@ -983,134 +1080,339 @@ void convert_to_row_annotator(const ColumnCompressed<std::string> &source,
-                    depth_terminal_count.fetch_add(1, std::memory_order_relaxed);
-                    continue;
+        }
+        after_chunk(chunk, succ_chunk.size());
+
+        progress_bar += succ_chunk.size();
+    }
+    assert(pred_boundary_it == pred_boundary.end());
+}
+
+
+void convert_batch_to_row_diff(const graph::DBGSuccinct &graph,
+                               const std::string &graph_fname,
+                               const std::vector<std::string> &source_files,
+                               const std::filesystem::path &dest_dir,
+                               uint32_t max_depth) {
+    if (source_files.empty())
+        return;
+
+    uint32_t num_threads = get_num_threads();
+    const bool build_terminal = build_successor(graph, graph_fname, max_depth, num_threads);
+
+    std::vector<std::unique_ptr<annot::ColumnCompressed<>>> sources;
+    for(const auto& fname : source_files) {
+        auto anno = std::make_unique<annot::ColumnCompressed<>>() ;
+        anno->merge_load({fname});
+        sources.push_back(std::move(anno));
+    }
+    logger->trace("Done loading {} annotations", sources.size());
+
+    sdsl::rrr_vector terminal;
+    std::ifstream f(graph_fname + ".terminal", std::ios::binary);
+    terminal.load(f);
+    f.close();
+
+    if (build_terminal) {
+        sdsl::bit_vector new_terminal(terminal.size());
+
+        // total number of set bits in the original and sparsified rows
+        std::vector<std::atomic<uint32_t>> orig_ones(chunk_size);
+        std::vector<std::atomic<uint32_t>> sparse_ones(chunk_size);
+
+        traverse_anno_chunked(
+                "Anchor opt", graph, graph_fname, sources,
+                [&]() {
+                    std::fill(orig_ones.begin(), orig_ones.end(), 0);
+                    std::fill(sparse_ones.begin(), sparse_ones.end(), 0);
+                },
+                [&](const bit_vector &source_col, uint64_t row_idx, uint64_t chunk,
+                    uint64_t, uint64_t, const std::vector<uint64_t> &succ_chunk,
+                    const std::vector<uint64_t> &, const std::vector<uint64_t> &) {
+                    if (terminal[row_idx])
+                        return;
+
+                    // if successor is not set, add it to the diff
+                    const uint64_t chunk_idx = row_idx - chunk;
+                    if (!source_col[succ_chunk[chunk_idx]])
+                        sparse_ones[row_idx]++;
+
+                    orig_ones[row_idx]++;
+                },
+                [&](uint64_t chunk_start, uint64_t chunk_size) {
+                    for (uint64_t i = 0; i < chunk_size; ++i) {
+                        if (sparse_ones[i] > 2 && sparse_ones[i] / 2 > orig_ones[i]) {
+                            new_terminal[i + chunk_start] = 1;
+                        } else {
+                            new_terminal[i + chunk_start] = terminal[i + chunk_start];
+                        }
+                    }
+                });
```

Comment:
use node_index instead of uint64_t where possible, so it's easier to read so see what refers to a node index and what is a simple chunk/file index